### PR TITLE
block existing bid season dates

### PR DESCRIPTION
**Pull request by elizabeth-jimenez**
Disable the dates on the date pickers that already have a bid season attached to them.

~[API PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/862)~

### CODE REVIEW

File: src/Components/AdministratorPage/ManageBidSeasons/EditBidSeasons.jsx [hunk 1]
```diff
@@ -60,6 +63,24 @@ const EditBidSeasons = (props) => {
...
+    delete bidSeasonDates$[formatDate(bidSeasonsBeginDate)];
+
+    const dateArr = [];
+    Object.values(bidSeasonDates$).forEach((seasonDates) => {
+      let loopDate = formatDate(seasonDates?.beginDate);
+      // eslint-disable-next-line no-loops/no-loops
+      while (isBefore(loopDate, formatDate(seasonDates?.endDate))) {
+        dateArr.push(new Date(loopDate));
+        loopDate = addDays(loopDate, 1);
+      }
```
Comment (elizabeth-jimenez) [inline src/Components/AdministratorPage/ManageBidSeasons/EditBidSeasons.jsx:78]: i could instead grab the number of days between the two dates and then forEach on them, to get rid of the while loop, but `while` feels more natural for this scenario

**Comment by elizabeth-jimenez** [comment]
Not very performant, so may not merge in - can keep here to play with optimization

File: src/Components/AdministratorPage/ManageBidSeasons/EditBidSeasons.jsx [hunk 2]
```diff
@@ -60,6 +63,24 @@ const EditBidSeasons = (props) => {
     swal.close();
   };
 
+  useEffect(() => {
+    const bidSeasonDates$ = JSON.parse(JSON.stringify(bidSeasonDates));
+    // remove current date range from bidSeasonDates$
```
Comment (elizabeth-jimenez) [inline src/Components/AdministratorPage/ManageBidSeasons/EditBidSeasons.jsx:68]: may be able to pull this out to the parent and generate it once, feeding custom dates based on the bid season range

**Comment by elizabeth-jimenez** [comment]
Scott noticed: they can still select a date before the start of a Bid Season all the way to after the end of that Bid Season - essentially completely engulfing a Bid Season

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx [hunk 3]
```diff
@@ -8,22 +8,30 @@ import { Column, Row } from 'Components/Layout';
 import { bidSeasonsEdit } from 'actions/BidSeasons';
 import EditBidSeasons from './EditBidSeasons';
 
-const ManageBidSeasonsCard = (props) => {
+// eslint-disable-next-line no-unused-vars
+const ManageBidSeasonsCard = ({ id, bidSeasonDateRanges, description, bidSeasonsBeginDate,
```
Comment (elizabeth-jimenez) [inline src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:12]: complaining about `id` not being used, even tho I do on L26

Comment (M00staff) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:12]: if you change L26 to `id={id}` you can remove the linter complaint

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx [hunk 4]
```diff
@@ -45,6 +45,41 @@ const ManageBidSeasons = () => {
...
+    // for new bid season - pull all date range dates
+    dateObj.allDates = Object.values(dateObj).flatMap(({ dates }) => dates);
+
+    // for edit bid season - pull all date range dates except current
+    Object.values(dateObj).forEach((seasonDate) => {
+      const currentBeginDate = formatDate(seasonDate?.beginDate);
+
+      dateObj[formatDate(seasonDate?.beginDate)].disableDates =
+        Object.values(dateObj).flatMap((seasonDateInner) =>
+          seasonDateInner?.beginDate === currentBeginDate ? [] : seasonDateInner?.dates);
```
Comment (M00staff) [inline src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:77]: The `.disableDates` will throw an error if `currentBeginDate` is null or undefined ... not sure yet why it's coming back that way

adding this logic in an if block fixes it - but i think there's something else off in this block

```
      if (currentBeginDate) {
        dateObj[currentBeginDate].disableDates =
          Object.values(dateObj).flatMap((seasonDateInner) =>
            seasonDateInner?.beginDate === currentBeginDate ? [] : seasonDateInner?.dates);
      }
```

image

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx [hunk 5]
```diff
@@ -8,22 +8,30 @@ import { Column, Row } from 'Components/Layout';
...
+          id
+          description
+          bidSeasonsBeginDate
+          bidSeasonsEndDate
+          bidSeasonsPanelCutoff
+          bidSeasonsFutureVacancy
           submitAction={submit}
-          {...props}
+          bidSeasonDisableDates={
+            bidSeasonDateRanges[formatDate(bidSeasonsBeginDate)]?.disableDates}
```
Comment (M00staff) [inline src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:34]: something wonky going on here as well - anytime i click on 'edit' on any card - i get the same dates & data populating the modal...

image

**Review by M00staff — CHANGES_REQUESTED** [review]
yea so this looks like a neat one to work on - but, a few things:

UX Stuff:
- when you open up 'add new bid season' - you just get entire months of un-pickable dates ... i think that might be super confusing for the user, especially without an error message or any kind of notification

- adding a button to change years on the datepicker would be nice for testing, but i dunno if that'd be useful to the user

- I can't tell that the disabled dates are disabled in Dark Mode

Code Stuff
- opening up the datepicker is really, really slow locally - i assume that lag would be much more pronounced in dev, and the other environments as well.  

- I think it's like you had mentioned - this logic is not very performant... it's also pretty gnarly to walk through and read.

- considering this logic is actually already handled in the database, I'm not sure why we would add it all a second time at such a big performance cost - when we can just expose the error message that comes back from the database, like we were doing - or at least just expand on that error handling logic

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx (hunk 3 above)
Comment (elizabeth-jimenez) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:12]: Ah thanks!

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx (hunk 5 above)
Comment (elizabeth-jimenez) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:34]: Unix epoch! Looks like it isnt receiving dates at all and i think i know why 😭

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx (hunk 4 above)
Comment (elizabeth-jimenez) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:77]: ugh so this is similar to the Agenda Item Validator error - after it was all working, i must have moved this line up `    dateObj.allDates = Object.values(dateObj).flatMap(({ dates }) => dates);` which broke it. Basically that loop expects to do the same modifications on the whole obj (and expects the same structure), but allDates deviates from all the other objs and doesn't need to be modified in that loop

Comment (elizabeth-jimenez) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasons.jsx:77]: was messing around with Object.seal and Object.freeze, but couldnt find a good use case to freeze top keys/values only

File: src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx (hunk 5 above)
Comment (elizabeth-jimenez) [reply src/Components/AdministratorPage/ManageBidSeasons/ManageBidSeasonsCard.jsx:34]: This is an error im realizing i make quite often and it's so sneaky! ur suggestion for id made me realize it. This was the fix 

```
          description={description}
          bidSeasonsBeginDate={bidSeasonsBeginDate}
          bidSeasonsEndDate={bidSeasonsEndDate}
          bidSeasonsPanelCutoff={bidSeasonsPanelCutoff}
          bidSeasonsFutureVacancy={bidSeasonsFutureVacancy}
          ```

**Comment by elizabeth-jimenez** [comment]
"UX Stuff:

when you open up 'add new bid season' - you just get entire months of un-pickable dates ... i think that might be super confusing for the user, especially without an error message or any kind of notification" Yea, meant to add an alert letting them know - added it, but left it more generic so it is ready to go after I delete all the disable dates stuff

"adding a button to change years on the datepicker would be nice for testing, but i dunno if that'd be useful to the user" oh good find - i didnt even notice we didnt have it, i know some of our date pickers do. Agreed, in this case i dont think it'd be useful.

"I can't tell that the disabled dates are disabled in Dark Mode" Oof! Nice catch

"Code Stuff

opening up the datepicker is really, really slow locally - i assume that lag would be much more pronounced in dev, and the other environments as well." Yea, I assume too 😬 

"I think it's like you had mentioned - this logic is not very performant... it's also pretty gnarly to walk through and read."
It is performant, like by the time u hit Edit it is all done, the lag is the library we use, so just wanted to get this to a place where we MAY be able to use it in the future after consolidating our date pickers (but im still going to delete it before we merge this cause i can just come back to this PR if we ever do)

"considering this logic is actually already handled in the database, I'm not sure why we would add it all a second time at such a big performance cost - when we can just expose the error message that comes back from the database, like we were doing - or at least just expand on that error handling logic" I think with another library this would be doable, just unfortunately not this one. The main thing i was trying to get away from is exposing all error messages to users without vetting them first. So I pushed a PR for the API to handle that and also added an alert to let users know they will get an error if it overlaps, before they ever make the call.

**Comment by M00staff** [comment]
yea - so the logic for all of this works... 

I'm curious though - how would the consolidating date picker ticket actually make this work better, if the library is the issue?

My understanding was - that ticket will have us create a 'shell' component that the existing DatePicker would live in, just so we can pass in props easier (which might not be entirely necessary, come to think of it 🤔) - not to create our own DatePicker, or pull in a different library to replace all the existing uses.

**Comment by elizabeth-jimenez** [comment]
> yea - so the logic for all of this works...
> 
> I'm curious though - how would the consolidating date picker ticket actually make this work better, if the library is the issue?
> 
> My understanding was - that ticket will have us create a 'shell' component that the existing DatePicker would live in, just so we can pass in props easier (which might not be entirely necessary, come to think of it 🤔) - not to create our own DatePicker, or pull in a different library to replace all the existing uses.

Cause we may consolidate our Date Pickers to one that works better with date disabling. We have about 3 different Date libraries we are using throughout TM and many more custom stylings strewn about for them - and every time someone has to work on implementing a date picker, it feels like a styling nightmare. "pull in a different library to replace all the existing uses" this is likely. That ticket is to standardize our Date stylings both for existing components and make it easier for future uses

**Comment by M00staff** [comment]
then we're definitely going to have to kick up the points on that ticket 🤔 ... by A LOT

**Pull request closed without merging**